auth_rust/src/main.rs

```diff
@@ -0,0 +1,244 @@
+    async fn acquire_token_value(&mut self, token: String) -> Result<TokenValue, AuthError> {
+        // Acquire redis value from the key from the `token` parameter
+        let token_value: RedisResult<Option<String>> = self.redis_client.get(token).await;
+
+        match token_value {
+            Ok(Some(value)) => {
+                let token_value: TokenValue = serde_json::from_str(&value).map_err(|_| AuthError::ParseError)?;
+                Ok(token_value)
+            }
+            Ok(None) => Err(AuthError::TokenNotRegistered),
+            Err(redis_error) => {
+                sentry::capture_error(&redis_error);
+                Err(AuthError::FailedToAcquireToken(redis_error))
+            },
+        }
+    }
+
+    #[tracing::instrument]
+    async fn acquire_counter_limit(&mut self, token_value: &TokenValue) -> Result<i64, AuthError> {
+        let formatted_date = chrono::Utc::now().format("%Y-%m").to_string();
+        let key = format!("counter/{}/{}", formatted_date, token_value.user_email);
+        let value: RedisResult<Option<String>> = self.redis_client.get(key).await;
+
+        match value {
+            Ok(Some(value)) => {
+                let counter_limit: i64 = value.parse().map_err(|_| AuthError::ParseError)?;
+                Ok(counter_limit)
+            }
+            Ok(None) => Ok(0),
+            Err(redis_error) => {
+                sentry::capture_error(&redis_error);
+                Err(AuthError::FailedToAcquireToken(redis_error))
+            },
+        }
+    }
+
+    #[tracing::instrument]
+    async fn increment_monthly_counter(
+        &mut self,
+        token_value: &TokenValue,
+        counter_limit: i64,
+    ) -> Result<(), AuthError> {
+        if token_value.user_email.starts_with("trial")
+            && token_value.user_email.ends_with("@pesto.teknologiumum.com")
+        {
+            return Ok(());
+        }
+
+        let formatted_date = chrono::Utc::now().format("%Y-%m").to_string();
+
+        let key = format!("counter/{}/{}", formatted_date, token_value.user_email);
+        let value = (counter_limit + 1).to_string();
+        let expiration: u64 = 60 * 60 * 24 * 40; // 40 days
+        let cmd: RedisResult<()> = self.redis_client.set_ex(key, value, expiration).await;
+        match cmd {
+            Ok(_) => Ok(()),
+            Err(redis_error) => {
+                sentry::capture_error(&redis_error);
+                Err(AuthError::FailedToAcquireToken(redis_error))
+            },
+        }
+    }
+}
+
+fn main() -> Result<(), Box<dyn Error>> {
+    tracing_subscriber::registry::Registry::default()
+        .with(sentry::integrations::tracing::layer())
+        .init();
+
+    let _guard = sentry::init((env::var("SENTRY_DSN").unwrap_or_default(), sentry::ClientOptions {
+        release: sentry::release_name!(),
+        sample_rate: 1.0,
+        traces_sample_rate: 0.5,
+        ..Default::default()
+    }));
+
+    tokio::runtime::Builder::new_multi_thread()
```

Comment:
We can use do 
```rust
#[tokio::main]
async fn main() -> _ { ... }
```
Instead of constructing tokio runtime ourself

Reply:
This causes some problem. See https://docs.sentry.io/platforms/rust/#async-main-function

Reply:
Mmm alright then